kimchi/src/circuits/cairo/runner/program.rs

```diff
@@ -0,0 +1,466 @@
+    pub mem: &'a mut CairoMemory<F>,
+    // comment instr for efficiency
+    /// current pointers
+    curr: CairoPointers<F>,
+    /// (if any) next pointers
+    next: Option<CairoPointers<F>>,
+    /// state auxiliary variables
+    vars: CairoVariables<F>,
+}
+
+/*
+/// Performs the addition of a u64 register with a signed offset
+fn add_off(reg: u64, off: i128) -> u64 {
+    // TODO(@querolita) check helper to avoid casting manually
+    // 1. Since the output address is u64, the offset cannot be larger than this
+    // 2. The absolute value of the offset will fit in u64
+    // 3. If it was negative, subtract. If it was positive, sum.
+    if off.is_negative() {
+        reg - (off.abs() as u64)
+    } else {
+        reg + (off.abs() as u64)
+    }
+}
+*/
+
+impl<'a, F: PrimeField> CairoStep<'a, F> {
+    /// Creates a new Cairo execution step from a step index, a Cairo word, and current pointers
+    pub fn new(mem: &mut CairoMemory<F>, ptrs: CairoPointers<F>) -> CairoStep<F> {
+        CairoStep {
+            //step: idx,
+            mem,
+            curr: ptrs,
+            next: None,
+            vars: CairoVariables::new(),
+        }
+    }
+
+    /// Executes a Cairo step from the current registers
+    pub fn execute(&mut self) {
+        // This order is important in order to allocate the memory in time
+        self.set_op0();
+        self.set_op1();
+        self.set_res();
+        self.set_dst();
+        // If the Option<> thing is not a guarantee for continuation of the program, we may be removing this
+        let next_pc = self.next_pc();
+        let (next_ap, next_fp) = self.next_apfp();
+        self.next = Some(CairoPointers::new(
+            next_pc.unwrap(),
+            next_ap.unwrap(),
+            next_fp.unwrap(),
+        ));
+    }
+
+    /// This function returns the current word instruction being executed
+    pub fn instr(&mut self) -> CairoWord<F> {
+        CairoWord::new(self.mem.read(self.curr.pc).unwrap())
+    }
+
+    /// This function computes the first operand address
+    pub fn set_op0(&mut self) {
+        if self.instr().op0_reg() == OP0_AP {
+            // reads first word from allocated memory
+            self.vars.op0_addr = self.curr.ap + self.instr().off_op0();
+        } else {
+            // reads first word from input parameters
+            self.vars.op0_addr = self.curr.fp + self.instr().off_op0();
+        } // no more values than 0 and 1 because op0_reg is one bit
+        self.vars.op0 = self.mem.read(self.vars.op0_addr);
+    }
+
+    /// This function computes the second operand address and content and the instruction size
+    pub fn set_op1(&mut self) {
+        if self.instr().op1_src() == OP1_DBL {
+            self.vars.size = F::one(); // double indexing
+            self.vars.op1_addr = self.vars.op0.unwrap() + self.instr().off_op1(); // should be positive for address
+            self.vars.op1 = self.mem.read(self.vars.op1_addr);
```

Comment:
Factor this out of all the branches?

Reply:
Ah well yes, I can put this aside (but taking into account the `else` branch, that corresponds to invalid instructions). But here again, if we were assuming that we are only executing the runner for valid bytecodes, then this should never arise.